### PR TITLE
[WIP] Add resource api spec

### PR DESCRIPTION
**Pull request by qiujian16**
https://issues.redhat.com/browse/OCM-4831

### CODE REVIEW

File: pkg/api/resource_types.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+func (d *Resource) BeforeCreate(tx *gorm.DB) error {
+	d.ID = NewID()
+	return nil
+}
+
+type DinosaurPatchRequest struct {
+	Species *string `json:"species,omitempty"`
+}
```
Comment (morvencao) [inline pkg/api/resource_types.go:29]: /DinosaurPatchRequest/ResourcePatchRequest/s

File: pkg/api/presenters/resource.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package presenters
+
+import (
+	"github.com/openshift-online/maestro/pkg/api"
+	"github.com/openshift-online/maestro/pkg/api/openapi"
+	"github.com/openshift-online/maestro/pkg/util"
+)
+
+func ConvertDinosaur(resource openapi.Resource) *api.Resource {
```
Comment (markturansky) [inline pkg/api/presenters/resource.go:9]: `ConvertResource`

File: pkg/api/presenters/resource.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	return &api.Resource{
+		Meta: api.Meta{
+			ID: util.NilToEmptyString(resource.Id),
+		},
+		ConsumerID: util.NilToEmptyString(resource.ConsumerId),
+		Version:    *resource.Version,
+	}
+}
+
+func PresentDinosaur(resource *api.Resource) openapi.Resource {
```
Comment (markturansky) [inline pkg/api/presenters/resource.go:19]: `PresentResource`

File: pkg/services/resource.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+		}
+		defer s.lockFactory.Unlock(ctx, lockOwnerID)
+	}
+
+	found, err := s.resourceDao.Get(ctx, dinosaur.ID)
+	if err != nil {
+		return nil, handleGetError("Resource", "id", dinosaur.ID, err)
+	}
+
+	// TODO version/consumer/status is immutable. Only object can be updated.
```
Comment (markturansky) [inline pkg/services/resource.go:110]: in a future PR, we can achieve this by including separating paths in openapi.yaml w/ an object that does not include status.

File: pkg/services/resource_test.go [hunk 5]
```diff
@@ -14,21 +14,21 @@ import (
 func TestDinosaurFindBySpecies(t *testing.T) {
```
Comment (markturansky) [inline pkg/services/resource_test.go:14]: Dinosaur

**Comment by markturansky** [comment]
You can add your service to the env, as seen here:

https://github.com/openshift-online/maestro/blob/main/cmd/maestro/environments/framework.go#L134

this is an area of improvement where we need a) documentation or better b) automatically integrate new services.

File: pkg/services/resource.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+var DisableAdvisoryLock = false
+
+type ResourceService interface {
+	Get(ctx context.Context, id string) (*api.Resource, *errors.ServiceError)
+	Create(ctx context.Context, dinosaur *api.Resource) (*api.Resource, *errors.ServiceError)
+	Replace(ctx context.Context, dinosaur *api.Resource) (*api.Resource, *errors.ServiceError)
+	Delete(ctx context.Context, id string) *errors.ServiceError
+	All(ctx context.Context) (api.ResourceList, *errors.ServiceError)
+
+	FindBySpecies(ctx context.Context, species string) (api.ResourceList, *errors.ServiceError)
```
Comment (clyang82) [inline pkg/services/resource.go:24]: Do we need this method? if so, may change it to `FindByTypes`

Comment (morvencao) [reply pkg/services/resource.go:24]: This is template code, we can drop that and add anyone useful, eg, `FindByConsumer` or `FindByCluster`.

**Comment by markturansky** [comment]
this is looking good overall. you know you're done when the tests pass :)  Until we get our automation in place, would you please cut and paste the output here for `make test && make test-integration`?

**Comment by qiujian16** [comment]
test passed

% make test-integration
CGO_ENABLED=1 GOEXPERIMENT=boringcrypto go install -ldflags="" ./cmd/maestro
OCM_ENV=testing gotestsum --format short-verbose -- -p 1 -ldflags -s -v -timeout 1h  \
			./test/integration
I1120 14:17:49.084541   99325 framework.go:72] Initializing testing environment
I1120 14:17:49.335392   99325 framework.go:151] Using Mock OCM Authz Client
I1120 14:17:49.335406   99325 framework.go:176] Disabling Sentry error reporting
I1120 14:17:49.372945   99325 api_server.go:145] Serving without TLS at localhost:8000
I1120 14:17:49.373058   99325 healthcheck_server.go:56] Serving HealthCheck without TLS at localhost:8083
PASS test/integration.TestControllerRacing (2.17s)
PASS test/integration.TestResourceGet (0.29s)
PASS test/integration.TestResourcePost (0.25s)
PASS test/integration.TestResourcePatch (0.28s)
PASS test/integration.TestResourcePaging (0.37s)
PASS test/integration.TestResourceListSearch (0.36s)
PASS test/integration.TestUpdateResourceWithRacingRequests (0.80s)
PASS test/integration.TestUpdateResourceWithRacingRequests_WithoutLock (0.73s)
I1120 14:17:54.662892   99325 api_server.go:151] Web server terminated

% make test
CGO_ENABLED=1 GOEXPERIMENT=boringcrypto go install -ldflags="" ./cmd/maestro
OCM_ENV=testing gotestsum --format short-verbose -- -p 1 -v  \
		./pkg/... \
		./cmd/...
EMPTY pkg/api
EMPTY pkg/api/openapi
EMPTY pkg/api/presenters
EMPTY pkg/auth
EMPTY pkg/client/ocm
PASS pkg/config.TestConfigReadStringFile (0.00s)
PASS pkg/config.TestConfigReadIntFile (0.00s)
PASS pkg/config.TestConfigReadBoolFile (0.00s)
PASS pkg/config.TestConfigReadQuotedFile (0.00s)
PASS pkg/config (cached)
PASS pkg/controllers.TestControllerFramework (0.00s)
PASS pkg/controllers (cached)
EMPTY pkg/dao
EMPTY pkg/dao/mocks
EMPTY pkg/db
EMPTY pkg/db/db_context
EMPTY pkg/db/db_session
EMPTY pkg/db/migrations
EMPTY pkg/db/mocks
EMPTY pkg/db/transaction
PASS pkg/errors.TestErrorFormatting (0.00s)
PASS pkg/errors.TestErrorFind (0.00s)
PASS pkg/errors (cached)
EMPTY pkg/handlers (cached)
EMPTY pkg/logger
PASS pkg/services.TestSQLTranslation (0.02s)
PASS pkg/services.TestResourceFindByConsumerID (0.00s)
PASS pkg/services
EMPTY pkg/util
EMPTY cmd/maestro
EMPTY cmd/maestro/clone
PASS cmd/maestro/environments.TestLoadServices (0.27s)
PASS cmd/maestro/environments (cached)
EMPTY cmd/maestro/migrate
EMPTY cmd/maestro/servecmd
EMPTY cmd/maestro/server
EMPTY cmd/maestro/server/logging

Note I have to disable the FK for consumer ID. Otherwise need to wait for #5 merged and rebase

**Review by clyang82 — APPROVED** [review]
/lgtm

**Pull request merged**